Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ILLLagrange.py

```diff
@@ -0,0 +1,308 @@
+
+    @staticmethod
+    def seeAlso():
+        return []
+
+    @staticmethod
+    def name():
+        return 'ILLLagrange'
+
+    def setup(self):
+        self.use_incident_energy = self.getProperty('UseIncidentEnergy').value
+        self.convert_to_wavenumber = self.getProperty('ConvertToWaveNumber').value
+
+    def PyInit(self):
+        self.declareProperty(MultipleFileProperty('SampleRuns', action=FileAction.Load, extensions=['']),
+                             doc='Sample run(s).')
+        self.declareProperty(MultipleFileProperty('ContainerRuns', action=FileAction.OptionalLoad, extensions=['']),
+                             doc='Container run(s) (empty cell)')
+        self.declareProperty(FileProperty('CorrectionFile', '', action=FileAction.OptionalLoad, extensions=['.txt']),
+                             doc='Correction reference file.')
+        self.declareProperty(MatrixWorkspaceProperty('OutputWorkspace', '', direction=Direction.Output),
+                             doc='The output workspace containing reduced data.')
+        self.declareProperty(name='UseIncidentEnergy', defaultValue=False,
+                             doc='Show the energies as incident energies, not transfer ones.')
+        self.declareProperty(name='ConvertToWaveNumber', defaultValue=False,
+                             doc='Convert axis unit to energy in wave number (cm-1)')
+
+        # the list of all the intermediate workspaces to group at the end
+        self.intermediate_workspaces = []
+
+    def PyExec(self):
+        self.setup()
+
+        correction_file = self.getPropertyValue('CorrectionFile')
+        self.output_ws_name = self.getPropertyValue('OutputWorkspace')
+
+        # load correction
+        if correction_file:
+            self.load_correction(correction_file)
+
+        empty_cell_files = self.getPropertyValue('ContainerRuns').split(',')
+
+        # empty cell treatment, if there is any
+        if empty_cell_files[0]:
+            # load and format empty cell
+            empty_cell_data = self.load_and_concatenate(empty_cell_files)
+            empty_cell_data = self.merge_adjacent_points(empty_cell_data)
+            energy, detector_counts, errors = self.get_counts_and_errors(empty_cell_data)
+
+            self.empty_cell_ws = "__" + self.output_ws_name + "_rawEC"
+
+            CreateWorkspace(outputWorkspace=self.empty_cell_ws,
+                            DataX=energy,
+                            DataY=detector_counts,
+                            DataE=errors,
+                            UnitX="Energy")
+
+            self.intermediate_workspaces.append(self.empty_cell_ws)
+
+            # correct by water
+            if self.water_correction is not None:
+                correctedECname = "__" + self.output_ws_name + "_waterCorrectedEC"
+                self.correct_data(self.empty_cell_ws, correctedECname)
+
+                self.intermediate_workspaces.append(correctedECname)
```

Comment:
Do you think the empty cell treatment could be extracted to a separate method to keep the PyExec as clean as possible?